src/blob/index/header.rs

```diff
@@ -0,0 +1,86 @@
+use super::prelude::*;
+
+#[derive(Debug, Deserialize, Serialize, Clone)]
+pub(crate) struct IndexHeader {
+    pub records_count: usize,
+    // contains serialized size of record headers, which allows to calculate offset in
+    // case of `OnDisk` state of indices
+    pub record_header_size: usize,
+    pub meta_size: usize,
+    pub hash: Vec<u8>,
+    // this field also contains `written` bit (the first one)
+    // to get the version, you should proceed `version >> 1`
+    pub(crate) version: u8,
+}
+
+impl IndexHeader {
+    pub fn new(record_header_size: usize, records_count: usize, meta_size: usize) -> Self {
+        Self {
+            records_count,
+            record_header_size,
+            meta_size,
+            ..Self::default()
+        }
+    }
+
+    pub fn with_hash(
+        record_header_size: usize,
+        records_count: usize,
+        meta_size: usize,
+        hash: Vec<u8>,
+    ) -> Self {
+        Self {
+            records_count,
+            record_header_size,
+            meta_size,
+            hash,
+            ..Self::default()
+        }
+    }
+
+    pub(crate) fn set_written(&mut self, state: bool) {
+        self.version = self.version & !1;
+        self.version |= state as u8;
```

Comment:
Please, recheck the conversion correctness. In C++ `true` `bool` value can be any non-negative number

Reply:
It is guaranteed [here](https://doc.rust-lang.org/std/primitive.bool.html), but if you worry that one day this could be changed I can match bool value and set 1 explicitly (otherwise resolve conversation please)